### PR TITLE
Reuse internal data as much as possible

### PR DESCRIPTION
**Pull request by furushchev**
The existing implementation is designed for computing all operations in a constructor method and we cannot reuse the allocated internal data for processing next coming images.
What is done in this PR are:
- Moved the operations done in constructor to `process` method
- Extracted allocations to `prealloc` from `process`

### CODE REVIEW

File: ED.cpp [hunk 1]
```diff
@@ -234,9 +266,11 @@ Mat ED::getGradImage()
   return result8UC1;
 }
 
-int ED::getSegmentNo() { return segmentNos; }
+Mat ED::getDirImage() { return dirImage; }
```
Comment (garaemon) [inline ED.cpp:269]: Return reference to reduce copy?

File: ED.cpp [hunk 2]
```diff
@@ -362,25 +398,27 @@ void ED::ComputeAnchorPoints()
...
   sortAnchorsByGradValue1(A);
+  const auto sort_anchors_by_grad_value_tick = getTickCount();
+  lastEDProfile.sort_anchors_by_grad_value = (sort_anchors_by_grad_value_tick - alloc_tick) / getTickFrequency();
 
   // Now join the anchors starting with the anchor having the greatest gradient value
   int totalPixels = 0;
+  int segmentNos = 0;
 
-  for (int k = anchorNos - 1; k >= 0; k--)
+  for (int k = anchorPoints.size() - 1; k >= 0; k--)
```
Comment (garaemon) [inline ED.cpp]: `static_cast<int>(anchorPoints.size())`

**Review by garaemon — APPROVED** [review]
Could you take a performance profile before and after this PR?

File: ED.cpp (hunk 1 above)
Comment (furushchev) [reply ED.cpp:269]: I think it is good to keep it as-is as a `cv::Mat` itself behaves like a smart pointer, the operation increments its reference counter but shares the same storage.

i.e.:

```
class Hoge                                                                                                                                                                   |class Hoge
{                                                                                                                                                                            |{
public:                                                                                                                                                                      |public:
 Hoge(const std::string& path) : m_image(cv::imread(path)) {}                                                                                                                | Hoge(const std::string& path) : m_image(cv::imread(path)) {}
 ~Hoge() {}                                                                                                                                                                  | ~Hoge() {}
                                                                                                                                                                             |
 cv::Mat getImage() { return m_image; }                                                                                                                                      | cv::Mat getImage() { return m_image; }
 cv::Mat& getImageRef() { return m_image; }                                                                                                                                  | cv::Mat& getImageRef() { return m_image; }
 cv::Mat m_image;                                                                                                                                                            | cv::Mat m_image;
};                                                                                                                                                                           |};
                                                                                                                                                                             |
int main(int argc, char** argv)                                                                                                                                              |int main(int argc, char** argv)
{                                                                                                                                                                            |{
  Hoge hoge("./billiard.jpg");                                                                                                                                               |  Hoge hoge("./billiard.jpg");
  printf("%p\n", hoge.m_image.ptr());                                                                                                                                        |  printf("%p\n", hoge.m_image.ptr());
  printf("%p\n", hoge.getImage().ptr());                                                                                                                                     |  printf("%p\n", hoge.getImage().ptr());
  printf("%p\n", hoge.getImageRef().ptr());                                                                                                                                  |  printf("%p\n", hoge.getImageRef().ptr());
  return 0;                                                                                                                                                                  |  return 0;
}                                                                                                                                                                            |}
```

outputs:

```
0x7f6119cff040
0x7f6119cff040
0x7f6119cff040
```
which indicates all mat point to the same data address.

Comment (garaemon) [reply ED.cpp:269]: OK, I understood.

File: ED.cpp (hunk 2 above)
Comment (furushchev) [reply ED.cpp]: Thank you! Changed.

**Comment by peroxyacyl** [comment]
@furushchev Any updates? I want to use this for s2cd from next week.

**Comment by furushchev** [comment]
@peroxyacyl Thank you for the reminder! I'll take a performance this week.

**Comment by peroxyacyl** [comment]
This patch enables s2 to read all four cameras at 0.3Hz. Previously it could not open in the first place.

**Comment by furushchev** [comment]
@peroxyacyl Thank you for the report and sorry that I could not have time to measure the change. I'll merge it not to block your experiment as for now.

**Pull request merged**